File: flutter_app/lib/domain/use_cases/prepare_acid_solution.dart
import '../../core/result.dart';
import '../entities/substance.dart';

class PrepareAcidSolution {
  const PrepareAcidSolution();

  Result<AcidPrepResult> call({
    required double stockPercent,
    required double stockDensity,
    required Substance acid,
    required double targetConcentration,
    required bool isMolarity,
    required double finalVolumeMl,
  }) {
    // Validation
    if (stockPercent <= 0 || stockPercent > 100) {
      return const Result.failure('Invalid stock concentration.');
    }
    if (finalVolumeMl <= 0) {
      return const Result.failure('Volume must be positive.');
    }
    if (acid.molecularWeight == null || acid.molecularWeight == 0) {
      return const Result.failure('Molecular weight missing for this substance.');
    }
    if (!isMolarity && (acid.basicity == null || acid.basicity == 0)) {
      return const Result.failure('Basicity missing for Normality calculation.');
    }

    try {
      final stockMolarity = (stockPercent * stockDensity * 10) / acid.molecularWeight!;
      final targetMolarity = isMolarity
          ? targetConcentration
          : targetConcentration / acid.basicity!.toDouble();

      if (targetMolarity > stockMolarity) {
        return const Result.failure('Target concentration exceeds stock.');
      }

      final volumeNeededMl = (targetMolarity * finalVolumeMl) / stockMolarity;

      return Result.success(
        AcidPrepResult(
          stockMolarity: stockMolarity,
          volumeNeededMl: volumeNeededMl,
          instructions: _generateInstructions(volumeNeededMl, finalVolumeMl, acid.name),
        ),
      );
    } catch (e) {
      return Result.failure('Calculation error: ${e.toString()}');
    }
  }

  String _generateInstructions(double vol, double finalVol, String acidName) =>
      'Measure ${vol.toStringAsFixed(1)} mL of $acidName. Add slowly to ~${(finalVol * 0.6).toInt()} mL water. Dilute to ${finalVol.toInt()} mL.';
}

File: web_app/lib/domain/prepareAcidSolution.ts
export type AcidPrepResult = {
  stockMolarity: number;
  volumeNeededMl: number;
  instructions: string;
};

export type Result<T> = 
  | { success: true; data: T }
  | { success: false; error: string };

export const prepareAcidSolution = (
  stockPercent: number,
  stockDensity: number,
  mw: number,
  basicity: number,
  targetConc: number,
  isMolarity: boolean,
  finalVolMl: number
): Result<AcidPrepResult> => {
  if (stockPercent <= 0 || stockPercent > 100) 
    return { success: false, error: 'Invalid stock concentration.' };
  
  const stockM = (stockPercent * stockDensity * 10) / mw;
  const targetM = isMolarity ? targetConc : targetConc / basicity;

  if (targetM > stockM)
    return { success: false, error: 'Target exceeds stock concentration.' };

  const vol = (targetM * finalVolMl) / stockM;
  return {
    success: true,
    data: {
      stockMolarity: stockM,
      volumeNeededMl: vol,
      instructions: `Measure ${vol.toFixed(1)} mL of acid. Add slowly to ~${(finalVolMl * 0.6).toFixed(0)} mL water. Dilute to ${finalVolMl.toFixed(0)} mL.`
    }
  };
};
